### ISSUE TITLE
Provide GitHub Desktop with the same Browse code feature as GitHub Mobile

### ISSUE THREAD

**Issue by nbinns** [desktop/desktop#13831]
I would like to browse my code in GitHub Desktop on Windows the same way I can in GitHub Mobile. Currently, it appears that the only way I can view my code via GitHub Desktop is through a third party entity, such as VS Code or Windows Explorer, both of which open up the possibility of accidentally editing the code, whereas a browse code feature would open the code in read-only mode.
The alternative is to log into the GitHub website, where code can be browsed, but that defeats the purpose and advantage of GitHub Desktop, as you have to present your login credentials each time you connect to GitHub via a browser.

It seems odd that this functionality doesn't already exist. Note, you can browse your commits and diffs in read-only mode directly through GitHub Desktop, but sometimes you just want to view the contents of a whole file. Thank you.

**Comment by 0xdevalias** [comment]
From a quick search in the code, it looks like both the `Changes` and `SelectedCommits` components seem to render `SeamlessDiffSwitcher`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/changes/changes.tsx#L99-L133

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/history/selected-commits.tsx#L163-L180

`SeamlessDiffSwitcher` is defined here:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/seamless-diff-switcher.tsx#L157-L165

And seems to use helper functions like `getFileContents`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/seamless-diff-switcher.tsx#L278-L281

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/syntax-highlighting/index.ts#L107-L130

Which calls `getOldFileContent` / `getNewFileContent`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/syntax-highlighting/index.ts#L107-L130

Which calls `getPartialBlobContents` / `readPartialFile`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/show.ts#L49-L83

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/file-system.ts#L55-L81

`SeamlessDiffSwitcher` eventually renders `Diff`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/seamless-diff-switcher.tsx#L351-L378

Which uses different render methods depending on the type:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/index.tsx#L121-L143

Of which `renderTextDiff` renders either `SideBySideDiff` or `TextDiff`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/index.tsx#L273-L309

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/side-by-side-diff.tsx#L215-L218

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/text-diff.tsx#L287-L292

From a quick skim through `SideBySideDiff`, it seems to do it's own unique UI rendering things, but `TextDiff` seems to wrap codemirror, which it renders with `CodeMirrorHost`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/text-diff.tsx#L640-L643

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/text-diff.tsx#L1550-L1575

`CodeMirrorHost` is defined here:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/diff/code-mirror-host.tsx#L92-L95

If you're not already aware, CodeMirror is a code editor component for the web.

- https://codemirror.net/

---

While there would obviously be other things to look into and consider in implementing this feature; the above code references should help point anyone who might be interested in looking into the feasibility of implementing this in the right sort of places in the code.

**Comment by tidy-dev** [comment]
Just a heads up, if a Desktop team member adds the `enhancement` label, that is not an acknowledgement of a feature we would like to see. It is just categorization of the issue as bug vs enhancement. GitHub Desktop does not currently have this type of functionality on the roadmap and would likely not accept a PR for it at this time. It is not inline with our goals as a Git client for GitHub (as opposed to a desktop version of github.com). If your goal is to get a PR merged, it is best to get a go ahead in the issue before investing a lot of work. Especially for a more complex feature, we like to see that the person developing the new feature has weighed the pros and cons and other potential solutions in the issue.

**Comment by 0xdevalias** [comment]
> GitHub Desktop does not currently have this type of functionality on the roadmap and would likely not accept a PR for it at this time.

@tidy-dev Yup, that's understandable. I was just exploring the issues and code and wanted to add relevant context as I had it.
